_addons_transformations/scale/readme.md

```diff
@@ -14,17 +14,47 @@ install: auto
 
 # Scale Transformation Service
 
-Transform the input by matching it between limits of ranges in a scale file.  The input string must be in numerical format.
+Transform a given input by matching it between limits of ranges.
+The input string must be in numerical format.
 
-The file is expected to exist in the `transform` directory, and should follow the format given in the example below.
+The file is expected to exist in the `transform` directory, and should follow the format given in the table below.
+
+| scale expression | returns XYZ when the given value is |
+|--------------------|--------|
+| `[12..23]=XYZ` | `between (or equal to) 12 and 23` |
+| `]12..23[=XYZ` | `between 12 and 23 (12 and 23 are excluded in this case.)` |
+| `..23]=XYZ` | `lower than or equal to 23` |
+| `]12..=XYZ` | `greater than 12` |
 
 ## Example
 
-transform/temps.scale:
+The following example shows how to break down numeric UV values into fixed UV index categories.
+We have an example UV sensor that sends us numeric values from `0` to `100`, which we then want to scale into the *UV Index* range.
+For further Information on the UV Index take a look at [Wikipedia](https://en.wikipedia.org/wiki/Ultraviolet_index).
+
+Example item:
+
+```java
+Number Uv_Sensor_Level "UV Level [SCALE(uvindex.scale):%s]"
+```
+
+Uvindex.scale in transform folder:
 
-```properties
-[12..23]=Temp between 12 or higher and 23 or lower
-]12..23[=Temp over 12 and below 23
-..23]=Any temp up to and including 23
-]12..=Any temp above 12
+```python
+..3]=1
+]3..6]=2
+]6..8]=3
+]8..10]=4
+[10..100]=5
 ```
+
+Each value the item receives, will be categorized to one of the five given ranges.
+Values **lower than or equal to 3** are catched with `..3]=1`.
+Greater values are catched in ranges with 2 values as criteria.
+The only condition here is that the received value has to be lower than or equal to `100` in our example, since we haven't defined other cases yet.
+If **none** of the configured conditions matches the given value, the response will be empty.
+
+Please note that all ranges for values above **3** are opened with a `]`.
+So the border values (3, 6, 8 and 10) are always transformed to the lower range, since the `]`excludes the given critera.
```

Comment:
That's not correct.

I have tested it with 
`[0..3]=1`
`[3..6]=2`
and
`[0..3]=1`
`]3..6]=2`
with an item value of `3`.

In the first constellation the item gets scaled to `2`.
Only with the *excluded* bracket, it will get scaled to `1`.

Reply:
Really? In this case I'd be interested in the result of

```
[3..6]=2
[0..3]=1
```

also

```

```
[0..7]=3
[3..6]=2
[0..3]=1
[0..7]=4
```

Reply:
Its `1` for your first example and `4` for the last one.

So it literally seems to walk through from top to bottom and uses the last matching range?!

Reply:
It seems like it but it is actually not that easy. I've taken a closer look at the source code and found that HashMap is used to store the parsed ranges. A HashMap does not have a guaranteed order. 

>  This class makes no guarantees as to the order of the map; in particular, it does not guarantee that the order will remain constant over time.

Can you confirm that it works multiple times? I'd very much like to guarantee a certain evaluation order in documentation. 😞 

@clinique you are the original developer of the SCALE transformation. Is the above a bug or am I missing something?

Reply:
@ThomDietrich :+1:  yes, we currently can not predict what range will be the first to intercept a given value. I would not say a bug. In my own usage, when I define scale file, I write them in order to have no overlap between ranges so I did not face this problem but in fact predictability in this case seems important, then it would be worth improving the scale transformation.

Reply:
Of course we could mention that the ranges have to be defined in order, but i think an improvement of the scale transformation would be the preferable approach.

Reply:
Ok, I can take care of it. Will had a specific testing to ensure we have a predictible evaluation order.

Reply:
@clinique Thanks for the quick response! I believe using LinkedHashMap would be a quick and painless solution. Additionally could you please reverse the Map to gain a top-to-bottom evaluation order? That would seem the most intuitive to me. Thanks for your consideration!

@martinvw all-clear 😄

Reply:
> Ok, I can take care of it. Will had a specific testing to ensure we have a predictible evaluation order.

@clinique did you already start, I was also on the way preparing something :-)

Reply:
@martinvw : I was about to go :)

Reply:
I also updated some doc-blocks and followed some suggestions of my idea and of the more recent coding guidelines. But I'm busy the rest of the day, so maybe you do the quick fix and I contribute my small changes later?

Reply:
Love the enthusiasm! @clinique While we are at it I had another look at the regex expression and found a few more things to fix or improve:

1. Other than claimed in the readme, bound symbols are not optional. The regex asks for either `[` or `]`. The expression `..23]` is not valid, `[..23]` would be valid. We should agree on one version or support both.
2. While we are at it, let's also support round brackets? https://en.wikipedia.org/wiki/Interval_(mathematics)#Classification_of_intervals
3. To catch another potential issue, do you think we should support and trim spaces in between elements?

Reply:
1. Yes, the contract as always be to defined range with '[' and ']'. They can not be omitted. The scale transformation relied on guava scale library when I wrote it. So was it.
2. Guava did not '('. From what I read in wikipedia, I think that adding parenthesis can be misleading for a poor value added as '(' <=> ']'. I do not think this is error prone.
3. Yes spaces should be trimed. I'll include some tests on this.

Reply:
@martinvw : the main problem is not generated by Hashmap but by Properties that is not ordered neither.

Reply:
Agreed, Accepted, Appreciated 😉 

@Confectrician you'll need to correct the README regarding 1.

Reply:
@clinique that makes it way harder, I might have time tonight, so please mention me on any pull request or if you will not be able to complete it.

Reply:
@ThomDietrich : regarding point 3, spaces can not be trimmed if we keep on using Properties to read scale definition file because they are not accepted in property key name and are then considered as a delimiter.

Reply:
I see. Weird... well then don't worry, we can do without

Reply:
@martinvw : I had to subclass Properties and use LinkedHashMap.